Change TypeId to be a thin wrapper around ConstantId.

zygoloid opened the pull request:
This better follows the principle that types are simply constants of type `type`, and allows more uniform treatment of types as just another kind of constant from generics handling.

Use a hash table to map from `TypeId` to information about the complete type. This makes basic operations on types a bit simpler, and operations that actually need to access the complete class information a bit more complex.

zygoloid committed:
Change TypeId to be a thin wrapper around ConstantId.

This better follows the principle that types are simply constants of
type `type`, and allows more uniform treatment of types as just another
kind of constant from generics handling.

Use a hash table to map from TypeId to information about the complete
type. This makes basic operations on types a bit simpler, and operations
that actually need to access the complete class information a bit more
complex.

jonmeow commented on toolchain/check/testdata/basics/no_prelude/multifile_raw_ir.carbon:
```diff
@@ -92,27 +94,29 @@ fn B() {
+// CHECK:STDOUT:     typeTypeType:    {kind: copy, type: typeTypeType}
+// CHECK:STDOUT:     typeError:       {kind: copy, type: typeError}
+// CHECK:STDOUT:     type7:           {kind: copy, type: type7}
+// CHECK:STDOUT:     type13:          {kind: none, type: type14}
+// CHECK:STDOUT:     type14:          {kind: none, type: type14}
+// CHECK:STDOUT:     type19:          {kind: none, type: type14}
```
Why does this end up with 3 references back to type14 (the function, I believe)? I see it's there before, so it's not a regression, just not sure why this happens.

zygoloid replied:
From the new output:
```
// CHECK:STDOUT:     'type(inst+4)':    {kind: none, type: type(inst+5)}
// CHECK:STDOUT:     'type(inst+5)':    {kind: none, type: type(inst+5)}
// CHECK:STDOUT:     'type(inst+10)':   {kind: none, type: type(inst+5)}
```
... which we can look up in the instructions list ...
```
// CHECK:STDOUT:     'inst+4':          {kind: FunctionType, arg0: function0, type: typeTypeType}
// CHECK:STDOUT:     'inst+5':          {kind: TupleType, arg0: type_block0, type: typeTypeType}
// CHECK:STDOUT:     'inst+10':         {kind: FunctionType, arg0: function1, type: typeTypeType}
```
... so these are two function types whose value representation is an empty tuple type, plus the empty tuple type itself.

jonmeow commented on toolchain/check/testdata/basics/no_prelude/multifile_raw_ir.carbon:
```diff
@@ -92,27 +94,29 @@ fn B() {
-// CHECK:STDOUT:     type3:           {constant: template inst+10, value_rep: {kind: none, type: type2}}
+// CHECK:STDOUT:     typeTypeType:    {kind: copy, type: typeTypeType}
+// CHECK:STDOUT:     typeError:       {kind: copy, type: typeError}
+// CHECK:STDOUT:     type7:           {kind: copy, type: type7}
```
Should this change how type numbers are printed to be offset by the builtin instruction (possibly ConstantId too)? Previously, we had:

```
// CHECK:STDOUT: type1: {constant: template inst+4, value_rep: {kind: none, type: type2}}
```

There, the constant `inst+4` is easy to look up:

```
// CHECK:STDOUT: 'inst+4': {kind: FunctionType, arg0: function0, type: typeTypeType}
```

Now, we have:

```
// CHECK:STDOUT: type7: {kind: copy, type: type7}
```

How do I figure out what `type7` is?

zygoloid replied:
Yeah, good point. I've changed printing for `TypeId` to basically print the `ConstantId`, so that you can see which instruction defines the type. For these lines we now produce:
```
// CHECK:STDOUT:     'type(instNamespaceType)': {kind: copy, type: type(instNamespaceType)}
// CHECK:STDOUT:     'type(inst+4)':    {kind: none, type: type(inst+5)}
// CHECK:STDOUT:     'type(inst+5)':    {kind: none, type: type(inst+5)}
// CHECK:STDOUT:     'type(inst+10)':   {kind: none, type: type(inst+5)}
```

zygoloid committed:
Change TypeId yaml output to say which constant the type represents.

Also add printing of symbolic constant IDs so that symbolic types can be
understood from the yaml output.

zygoloid committed:
Update based on @jonmeow's feedback.

jonmeow reviewed (approved):
Thanks!

zygoloid committed:
Merge branch 'trunk' into toolchain-type

zygoloid committed:
Fix non-auto test expectations.
